File: web/src/currencies.js
/**
 * The currencies monori knows about.
 *
 * A short curated list, not all of ISO 4217, and every entry has two minor
 * units — which is what lets every amount stay an integer count of them and the
 * currency be a label on top. `server/app/currencies.py` is the same list on
 * the server; a code offered here but rejected there would only ever surface as
 * a 400 in someone's account dialog, so `test_currencies.py` pins them together.
 */
export const CURRENCIES = [
    { code: "RUB", name: "Russian ruble", symbol: "₽" },
    { code: "USD", name: "US dollar", symbol: "$" },
    { code: "EUR", name: "Euro", symbol: "€" },
    { code: "GBP", name: "Pound sterling", symbol: "£" },
    { code: "CHF", name: "Swiss franc", symbol: "CHF" },
    { code: "KZT", name: "Kazakhstani tenge", symbol: "₸" },
    { code: "BYN", name: "Belarusian ruble", symbol: "Br" },
    { code: "GEL", name: "Georgian lari", symbol: "₾" },
    { code: "AMD", name: "Armenian dram", symbol: "֏" },
    { code: "TRY", name: "Turkish lira", symbol: "₺" },
    { code: "AED", name: "UAE dirham", symbol: "AED" },
    { code: "CNY", name: "Chinese yuan", symbol: "¥" },
    { code: "RSD", name: "Serbian dinar", symbol: "RSD" },
];

export const DEFAULT_CURRENCY = "RUB";

const BY_CODE = new Map(CURRENCIES.map((c) => [c.code, c]));

/** " gel " -> "GEL"; anything blank becomes the fallback */
export function normalizeCurrency(code, fallback = DEFAULT_CURRENCY) {
    const normalized = String(code ?? "")
        .trim()
        .toUpperCase();
    return normalized || fallback;
}

/** The symbol printed after an amount; an unknown code prints as itself. */
export function currencySymbol(code) {
    const normalized = normalizeCurrency(code, "");
    return BY_CODE.get(normalized)?.symbol ?? normalized;
}

export function currencyName(code) {
    const normalized = normalizeCurrency(code, "");
    return BY_CODE.get(normalized)?.name ?? normalized;
}

/** Options for a select; an unknown code (older account, imported data) is kept
 * as its own option so editing an account never silently rewrites it. */
export function currencyOptions(current) {
    const options = CURRENCIES.map((c) => ({ value: c.code, label: `${c.code} · ${c.name}` }));
    const code = normalizeCurrency(current, "");
    if (code && !BY_CODE.has(code)) {
        options.unshift({ value: code, label: code });
    }
    return options;
}

File: server/app/currencies.py
"""
The currencies monori knows about.

Deliberately a short, curated list rather than all of ISO 4217: every code here
is one a monori user plausibly holds money in, and every one of them has two
minor units. That second property is what lets the whole codebase keep storing
money as an integer count of minor units ("kopecks") and treat the currency as a
label on top — no per-currency scaling anywhere.

``web/src/currencies.js`` is the same list on the frontend; the two are kept in
step by ``tests/unit/test_currencies.py``.
"""

from fastapi import HTTPException

MINOR_UNITS = 2

# code -> (english name, symbol)
CURRENCIES = {
    "RUB": ("Russian ruble", "₽"),
    "USD": ("US dollar", "$"),
    "EUR": ("Euro", "€"),
    "GBP": ("Pound sterling", "£"),
    "CHF": ("Swiss franc", "CHF"),
    "KZT": ("Kazakhstani tenge", "₸"),
    "BYN": ("Belarusian ruble", "Br"),
    "GEL": ("Georgian lari", "₾"),
    "AMD": ("Armenian dram", "֏"),
    "TRY": ("Turkish lira", "₺"),
    "AED": ("UAE dirham", "AED"),
    "CNY": ("Chinese yuan", "¥"),
    "RSD": ("Serbian dinar", "RSD"),
}

DEFAULT_CURRENCY = "RUB"

# every rate is quoted against this one, so a conversion is always two hops at
# most and the rate table stays one row per currency per day
PIVOT_CURRENCY = "RUB"


def normalize(code, fallback=DEFAULT_CURRENCY):
    """
    Trim and upper-case a currency code, falling back when it is missing.
    Unknown codes are returned as given — validation is a separate decision, so
    data that predates the registry can still be read back.
    """
    code = (code or "").strip().upper()
    return code or fallback


def is_known(code):
    return normalize(code, "") in CURRENCIES


def validate(code, fallback=DEFAULT_CURRENCY):
    """
    Normalize a code coming from a request, rejecting anything not on the list.
    """
    normalized = normalize(code, fallback)
    if normalized not in CURRENCIES:
        raise HTTPException(400, f"unknown currency {normalized!r}")
    return normalized


def symbol(code):
    entry = CURRENCIES.get(normalize(code, ""))
    return entry[1] if entry else normalize(code, "")


def catalog():
    """
    The registry as the API serves it.
    """
    return [
        {"code": code, "name": name, "symbol": sym, "minorUnits": MINOR_UNITS}
        for code, (name, sym) in CURRENCIES.items()
    ]
